p2p/socket.go

```diff
@@ -0,0 +1,461 @@
+package p2p
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/pokt-network/pocket/p2p/types"
+
+	"go.uber.org/atomic"
+)
+
+type SocketKind string
+
+var (
+	Outbound            SocketKind = "outbound"
+	Inbound             SocketKind = "inbound"
+	UndefinedSocketKind SocketKind = "unspecified"
+)
+
+type SocketEventMonitor func(context.Context, *socket) error
+
+// socket is an abstraction around net.Conn to represent a p2p connection with full read/write capabilities
+// both read and write operations are buffer, and both buffer sizes are configurable through configuration
+// configuration paramters are directly assigned to the socket struct
+// 1 live p2p connection = 1 socket
+type socket struct {
+	// the agent responsible for running/creating/managing sockets
+	runner types.Runner
+
+	// configuration parameters
+	headerLength uint
+	bufferSize   uint
+	readTimeout  uint
+	addr         string
+	kind         SocketKind // inbound or outbound
+
+	// the actual network socket
+	conn net.Conn
+
+	// the wire codec
+	c *wireCodec
+
+	// io buffers
+	buffers struct {
+		read  *types.Buffer
+		write *types.ConcurrentBuffer
+	}
+
+	// the io reader/writer
+	reader *bufio.Reader
+	writer *bufio.Writer
+
+	// the map to track writes that expects acknowledgements
+	// we call them requests (as they require responses)
+	requests *types.RequestMap
+
+	// turns true when the socket is opened (i.e: the connection is established and IO is on going)
+	isOpen atomic.Bool
+
+	ready   chan uint // when the socket is opened and IO starts, this channel gets closed to signal readiness
+	writing chan uint // when the writing starts, this channel receives a new input; closes when done writing (i.e: stopped the socket)
+	reading chan uint // when the reading starts, this channel receives a new input; closes when done reading (i.e: stopped the socket)
+	done    chan uint // if this channel is closed or receives and input, it stops the socket and IO operations
+	closed  chan uint // this channel signals that the socket has been closed by receiving a new input
+
+	errored chan uint // on error, this channel receives a new input to signal the happening of an error
+	err     struct {  // the reference to store the encountered error.
```

Comment:
Yeah... as soon as it errors it stops, for the time being all errors that this struct stores are halting. This can definitely use some more thinking as telemetry and error handling are topics I haven't dealt with properly yet.